### PR TITLE
Implement better progress

### PR DESCRIPTION
**Pull request by iluuu1994**
The purpose of this merge request:

* Errors should be written to `STDERR`
    * This allows piping the output to other commands (`psalm --output-format=json > result.json`) without having to worry about writing things to the screen
* Make it possible to implement other progress bars

### CODE REVIEW

**Comment by iluuu1994** [comment]
@muglug This contains lots or BC breaks (removal of `$debug_output`). Is this even feasible? Do we need a different solution without breaking BC?

**Comment by muglug** [comment]
Pretty sure nobody uses `$debug_output` outside of Psalm, so feel free!

How would a hypothetical future progress bar deal with threaded execution.

**Comment by iluuu1994** [comment]
> How would a hypothetical future progress bar deal with threaded execution.

Good point. Forking makes sharing state very difficult. But I'm not sure rewriting everything to theads is worth it. I'll see if there a simple solution.

**Comment by iluuu1994** [comment]
Wasn't aware you're already using sockets to communicate between the processes. Shouldn't be too hard to extend 🙂

**Comment by iluuu1994** [comment]
Here's a simple [POC](https://github.com/iluuu1994/psalm/compare/feature/progress...iluuu1994:feature/progress-poc). Basically, the serialized message is `base64_encode`d and the messages are separated by `PHP_EOL`. This way the socket can send an indefinite amount of messages. I'll clean this up and create a proper progress bar soon.

**Comment by muglug** [comment]
Looking v. promising!

**Comment by iluuu1994** [comment]
image

@muglug

* Would you like to keep the "Scanning files..." and "Analyzing files..." messages? 
* The progress for scanning files is currently skipped because it's pretty fast

**Comment by erunion** [comment]
This is rad!

**Comment by iluuu1994** [comment]
Not necessarily perfect but I'd like to hear your feedback 🙂

File: src/Psalm/Internal/Codebase/Analyzer.php [hunk 1]
```diff
@@ -230,13 +226,17 @@ public function analyzeFiles(ProjectAnalyzer $project_analyzer, $pool_size, $alt
...
-                }
+                $this->progress->debug('Analyzing ' . $file_analyzer->getFilePath() . "\n");
 
                 $file_analyzer->analyze(null);
             };
 
+        $this->progress->start(count($this->files_to_analyze));
+
+        $task_done_closure = function (): void {
+            $this->progress->taskDone(true);
```
Comment (iluuu1994) [inline src/Psalm/Internal/Codebase/Analyzer.php]: Completely forgot about this part 😆

**Comment by muglug** [comment]
This is fantastic, thanks!

**Comment by muglug** [comment]
The only thing I'm vaguely bothered by is the cosmetic similarity between this and PHPUnit output - when scanning a long build output, I worry the two might be confused.

I wonder whether there's some way of formatting the output to look a _little_ different?

Here's what they look like now:

```
$ ./psalm --threads=8
Scanning files...
Analyzing files...
............................................................  60 / 650 (9%)
............................................................ 120 / 650 (18%)
............................................................ 180 / 650 (28%)
............................................................ 240 / 650 (37%)
............................................................ 300 / 650 (46%)
............................................................ 360 / 650 (55%)
............................................................ 420 / 650 (65%)
............................................................ 480 / 650 (74%)
............................................................ 540 / 650 (83%)
............................................................ 600 / 650 (92%)
..................................................

------------------------------
No errors found!
------------------------------

Checks took 8.72 seconds and used 440.572MB of memory
Psalm was able to infer types for 99.7917% of the codebase

$ vendor/bin/phpunit
PHPUnit 8.1.5 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.3.5
Configuration: /Users/matthewbrown/Desktop/vimeo/git/psalm/phpunit.xml.dist
Error:         No code coverage driver is available

.............................................................   61 / 3383 (  1%)
.............................................................  122 / 3383 (  3%)
.............................................................  183 / 3383 (  5%)
.............................................................  244 / 3383 (  7%)
.............................................................  305 / 3383 (  9%)
.............................................................  366 / 3383 ( 10%)
.............................................................  427 / 3383 ( 12%)
.............................S..S............................  488 / 3383 ( 14%)
.............................................................  549 / 3383 ( 16%)
.............................................................  610 / 3383 ( 18%)
.............................................................  671 / 3383 ( 19%)
.............................................................  732 / 3383 ( 21%)
```

**Comment by muglug** [comment]
What about

```
./psalm --threads=8
Scanning files...
Analyzing files...
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  60 / 650 (9%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 120 / 650 (18%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 180 / 650 (28%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 240 / 650 (37%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 300 / 650 (46%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 360 / 650 (55%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 420 / 650 (65%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 480 / 650 (74%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 540 / 650 (83%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 600 / 650 (92%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
```

**Comment by weirdan** [comment]
That assumes Unicode output (depending on console emulator, font and output codepage it may produce vastly different results). For compatibility reasons I would suggest `*`, `#`, `*`, `-` or `+` instead (my personal preference being `#` perhaps).

```
./psalm --threads=8
Scanning files...
Analyzing files...
############################################################  60 / 650 (9%)
############################################################ 120 / 650 (18%)
############################################################ 180 / 650 (28%)
############################################################ 240 / 650 (37%)
############################################################ 300 / 650 (46%)
############################################################ 360 / 650 (55%)
############################################################ 420 / 650 (65%)
############################################################ 480 / 650 (74%)
############################################################ 540 / 650 (83%)
############################################################ 600 / 650 (92%)
##############################################
```

Light shade (the character you proposed) looks like this 
image

**Comment by iluuu1994** [comment]
@muglug Awesome! Thanks for merging!

> I wonder whether there's some way of formatting the output to look a little different?

It's your call, really. The output is also used by [PHP-CS-Fixer](https://github.com/FriendsOfPHP/PHP-CS-Fixer), [PHP_CodeSniffer](https://github.com/squizlabs/PHP_CodeSniffer) and probably more so I've gotten kinda used to it.

@weirdan 

> my personal preference being # perhaps

The problem is that it adds a lot of noise, making failures (`F`) less pronounced.

**Comment by muglug** [comment]
Sorry - I had to revert this, as forking isn't working properly anymore - `./psalm --threads=8 --no-cache` fails for me locally. I think it's not creating a proper clone of the `FileStorageProvider` anymore.

**Comment by iluuu1994** [comment]
@muglug Bummer.

`./psalm --threads=8 --no-cache` works fine on my last commit (5e8acf0b) but it fails on `master`. Looks like something else causes this.

**Comment by iluuu1994** [comment]
04ef20da1fb98d8ec1b3d0d7474dbdb357f9a2f5 introduces the problem apparently.

**Comment by iluuu1994** [comment]
Jumped the gun. It is my PR, it just happens like every 5th time. I'll investigate.

**Comment by muglug** [comment]
Thanks! It’s a really great addition.

**Comment by iluuu1994** [comment]
This is super weird. Sometimes the parent does not receive the entire message before `feof` is true. However, the forked process does not fail. I found [this](https://stackoverflow.com/a/15710545/1320374) that says:

> `feof` might return `true` (socket is closed) while PHP still has some data in its buffer.

but even getting rid of `feof` and checking if `fread` returns `false` does not solve the problem.

**Comment by muglug** [comment]
Maybe an interim solution is to disable the progressbar when forking, with a suitable message given to the user during threaded execution?

**Comment by iluuu1994** [comment]
I'm sure I'll find a solution, just need a bit more time 🙂 I'll have another look this Thursday at the latest.

File: src/Psalm/Internal/Fork/Pool.php [hunk 2]
```diff
@@ -227,6 +240,23 @@ private function readResultsFromChildren()
...
+                        }
+                    } else {
+                        error_log('Child should return ForkMessage - response type=' . gettype($message));
+                        $this->did_have_error = true;
+                    }
+                }
+
                 // If the stream has closed, stop trying to select on it.
                 if (feof($file)) {
                     fclose($file);
```
Comment (weirdan) [inline src/Psalm/Internal/Fork/Pool.php]: I would suggest to assert that `$content[intval($file)]` is empty here.

File: src/Psalm/Internal/Fork/Pool.php [hunk 3]
```diff
@@ -227,6 +240,23 @@ private function readResultsFromChildren()
                     $content[intval($file)] .= $buffer;
                 }
 
+                if (strpos($buffer, PHP_EOL) === strlen($buffer) - 1) {
```
[the remark this answers is not in the document]
Comment (iluuu1994) [reply src/Psalm/Internal/Fork/Pool.php]: I thought about that too but wasn't sure if that was possible. I don't think it's the culprit but I should probably still adjust it.

File: src/Psalm/Internal/Fork/Pool.php [hunk 4]
```diff
@@ -227,6 +240,23 @@ private function readResultsFromChildren()
                     $content[intval($file)] .= $buffer;
```
[the remark this answers is not in the document]
Comment (iluuu1994) [reply src/Psalm/Internal/Fork/Pool.php]: Good point.

File: src/Psalm/Internal/Fork/Pool.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (iluuu1994) [reply src/Psalm/Internal/Fork/Pool.php]: Turns out this solved the problem after all, thanks @weirdan!

**Pull request merged**